dascore/proc/correlate.py

```diff
@@ -3,143 +3,204 @@
+    of float.
+
+    Parameters
+    ----------
+    patch
+        The input patch
+    dim
+        The dimension name that was correlated in the freq. domain.
+    undo_weighting
+        If True, also undo the weighting artifact caused by DASCore's dft
+        weighting. This is done by simply dividing by the coordinate step.
+        See [dft note](`docs/notes/dft_notes.qmd`) for more details.
+
+    Examples
+    --------
+    >>> import dascore as dc
+    >>> patch = dc.get_example_patch()
+    >>>
+    >>> # Example 1
+    >>> # An auto-correlation of the example patch
+    >>> dft = patch.dft("time", real=True)
+    >>> dft_sq = dft * dft.conj()
+    >>> idft = dft_sq.idft()
+    >>> auto_patch = idft.correlate_shift(dim="time")
     """
-    ndims = len(data.shape)
-    start_slice = slice(-np.floor(cx_len / 2).astype(int), None)
-    start_ind = broadcast_for_index(ndims, axis, start_slice)
-    stop_slice = slice(None, np.ceil(cx_len / 2).astype(int))
-    stop_ind = broadcast_for_index(ndims, axis, stop_slice)
-    data1 = data[start_ind]
-    data2 = data[stop_ind]
-    data = np.concatenate([data1, data2], axis=axis)
-    return data
+    coord = patch.get_coord(dim, require_evenly_sampled=True)
+    axis = patch.dims.index(dim)
+    data = np.fft.fftshift(patch.data, axes=axis)
+    if undo_weighting:
+        data = data / to_float(coord.step)
+    # so it appears (from testing) there is one lest sample on the positive
+    # side.
+    step = to_float(coord.step)
+    new_start = -np.ceil((len(coord) - 1) / 2) * step
+    new_end = np.ceil((len(coord) - 1) / 2) * step
+    new_coord = dc.get_coord(start=new_start, stop=new_end, step=step)
+    assert len(new_coord) == len(coord)
+    cm = patch.coords
+    new_cm = cm.update(**{dim: new_coord}).rename_coord(**{dim: f"lag_{dim}"})
+    out = patch.update(data=data, coords=new_cm)
+    return out
 
 
 @patch_function()
 def correlate(
-    patch: PatchType, lag: int | float | Quantity | None = None, samples=False, **kwargs
+    patch: PatchType,
+    samples=False,
+    **kwargs,
 ) -> PatchType:
     """
-    Correlate a single row/column in a 2D patch with every other row/column.
+    Correlate source row/columns in a 2D patch with all other row/columns.
+
+    Correlations are done in the frequency domain. This function can accept a
+    patch whose target dimension has already been transformed with the
+    [`Patch.dft`](`dascore.transform.fourier.dft`) method, otherwise the dft
+    will be performed. If the input has already been transformed,
+    [`Patch.correlation_shift`](`dascore.proc.correlate.correlate_shift`)
+    is useful to undo dft artefacts after the idft is applied.
+
+    While a 2D patch is required for input, a 3D patch is returned where the
+    3rd dimesion cooresponds to the source rows/columns. For the case of a
+    single source, the [`Patch.squeeze`](`dascore.Patch.squeeze`) method
+    can be helpful to remove length 1 dimensions.
```

Comment:
I suggest we squeeze under the hood for a single source case so the patch is already 2D as expected.

Reply:
I went back and forth on this. 

On the one hand, it is easier to deal with a 2D patch. On the other though, having the third dimension preserves the information about which channel was used as the source, then the user gets to choose if they want to throw this away with a squeeze or not.

Reply:
Sounds good!